rasa/nlu/utils/hugging_face/hf_transformers.py

```diff
@@ -0,0 +1,328 @@
+            )
+        return actual_sequence_lengths, padded_token_ids
+
+    @staticmethod
+    def _extract_nonpadded_embeddings(embeddings, actual_sequence_lengths):
+
+        nonpadded_sequence_embeddings = []
+        for index, embedding in enumerate(embeddings):
+            unmasked_embedding = embedding[: actual_sequence_lengths[index]]
+            nonpadded_sequence_embeddings.append(unmasked_embedding)
+
+        return np.array(nonpadded_sequence_embeddings)
+
+    def _compute_batch_sequence_features(self, batch_attention_mask, padded_token_ids):
+
+        model_outputs = self.model(
+            np.array(padded_token_ids), attention_mask=np.array(batch_attention_mask)
+        )
+
+        # sequence hidden states is always the first output from all models
+        sequence_hidden_states = model_outputs[0]
+
+        sequence_hidden_states = sequence_hidden_states.numpy()
+        return sequence_hidden_states
+
+    def _get_model_features_for_batch(
+        self, batch_token_ids: List[List[int]]
+    ) -> np.array:
+
+        # Let's first add tokenizer specific special tokens to all examples
+        batch_token_ids_augmented = self._add_lm_specific_special_tokens(
+            batch_token_ids
+        )
+
+        # Let's first add padding so that whole batch can be fed to the model
+        actual_sequence_lengths, padded_token_ids = self._add_padding_to_batch(
+            batch_token_ids_augmented
+        )
+
+        # Compute attention mask based on actual_sequence_length
+        batch_attention_mask = self._compute_attention_mask(actual_sequence_lengths)
+
+        # Get token level features from the model
+        sequence_hidden_states = self._compute_batch_sequence_features(
+            batch_attention_mask, padded_token_ids
+        )
+
+        # Extract features for only non-padding tokens
+        sequence_nonpadded_embeddings = self._extract_nonpadded_embeddings(
+            sequence_hidden_states, actual_sequence_lengths
+        )
+
+        # Extract sentence level and post-processed features
+        (
+            sentence_embeddings,
+            sequence_final_embeddings,
+        ) = self._post_process_sequence_embeddings(sequence_nonpadded_embeddings)
+
+        return sentence_embeddings, sequence_final_embeddings
+
+    def _get_docs_for_batch(
+        self, batch_examples: List[Message], attribute: Text
+    ) -> List[Dict[Text, Any]]:
+
+        batch_tokens, batch_token_ids = self._get_token_ids_for_batch(
+            batch_examples, attribute
+        )
+
+        (
+            batch_sentence_features,
+            batch_sequence_features,
+        ) = self._get_model_features_for_batch(batch_token_ids)
+
+        # A doc consists of {'token_ids': ..., 'tokens': ..., 'sequence_features': ..., 'sentence_features': ...}
+        batch_docs = []
+        for index in range(len(batch_examples)):
+            doc = {
+                "token_ids": batch_token_ids[index],
```

Comment:
Not sure, but should we use constants for the keys? They could be defined in this class and then reused everywhere else. Prevents typing errors and allows us to easily rename things.

Reply:
Added constants